src/module.c

```diff
@@ -504,6 +502,34 @@ void *RM_PoolAlloc(RedisModuleCtx *ctx, size_t bytes) {
  * Helpers for modules API implementation
  * -------------------------------------------------------------------------- */
 
+client *moduleAllocTempClient() {
+    client *c;
+
+    pthread_mutex_lock(&moduleTempClientsMutex);
+    if (moduleTempClientCount > 0) {
+        c = moduleTempClients[--moduleTempClientCount];
+    } else {
+        c = createClient(NULL);
+        c->flags |= CLIENT_MODULE;
+        c->user = NULL; /* Root user */
+    }
+    pthread_mutex_unlock(&moduleTempClientsMutex);
+    return c;
+}
+
+void moduleReleaseTempClient(client *c) {
+    pthread_mutex_lock(&moduleTempClientsMutex);
+    if (moduleTempClientCount == MODULE_MAX_TEMP_CLIENT_COUNT) {
+        freeClient(c);
```

Comment:
Looking at `freeClient`, it may involve quite a few different flows and may even give control back to modules. I think it should only be called with the mutex unlocked.

Reply:
Looks like clearClient() may also call back to modules. createClient() doesn't but it may in future. Changed code a bit, holding the lock just to add/remove to/from array.